logproof/src/inner_product.rs

```diff
@@ -655,5 +701,76 @@ mod tests {
         fn can_verify_valid_proof_11() {
             validate_proof(11);
         }
+
+        #[test]
+        fn can_msm_generator_folding() {
+            let gens = LogProofGenerators::new(8);
+
+            let c = (0..3)
+                .map(|_| Scalar::random(&mut thread_rng()))
+                .collect::<Vec<_>>();
+
+            let get_expected_gens = || -> (RistrettoPoint, RistrettoPoint) {
+                let mut g = gens.g.clone();
+                let mut h = gens.h.clone();
+                let mut i = 0;
+
+                loop {
+                    let n_2 = g.len() / 2;
+
+                    if n_2 == 0 {
+                        assert_eq!(g.len(), 1);
+                        return (g[0], h[0]);
+                    }
+
+                    let (g_l, g_r) = g.split_at(n_2);
+                    let (h_l, h_r) = h.split_at(n_2);
+
+                    let g_l = RistrettoPointVec::new(g_l);
+                    let g_r = RistrettoPointVec::new(g_r);
+                    let h_l = RistrettoPointVec::new(h_l);
+                    let h_r = RistrettoPointVec::new(h_r);
+
+                    let c = c[i];
+                    let c_inv = c.invert();
+
+                    g = (g_l + g_r * c).into_iter().collect::<Vec<_>>();
+                    h = (h_l + h_r * c_inv).into_iter().collect::<Vec<_>>();
+                    i += 1;
+                }
+            };
+
+            let s_i = |i| {
+                c.iter().rev().enumerate().fold(Scalar::one(), |p, (j, x)| {
+                    if i & (0x1 << j) != 0 {
+                        p * x
+                    } else {
+                        p
+                    }
+                })
+            };
```

Comment:
We should probably factor out the common parts that are duplicated here, so that the test is always testing the actual implementation.

Reply:
This test can actually be removed. It was a TDD thing I made to test that the algorithm for computing the s values worked. We now know the algorithm works because verification tests still pass.